tests/522_explicit_struct_names.rs

```diff
@@ -0,0 +1,62 @@
+use ron::{extensions::Extensions, from_str, Error, Options};
+use serde_derive::Deserialize;
+
+#[derive(Debug, Deserialize)]
+struct Id(u32);
+
+#[derive(Debug, Deserialize)]
+struct Position(f32, f32);
+
+#[derive(Debug, Deserialize)]
+struct Foo {
+    #[allow(unused)]
+    pub id: Id,
+    #[allow(unused)]
+    pub position: Position,
+}
+
+const EXPECT_ERROR_MESSAGE: &'static str =
+    "expected `Err(Error::ExpectedStructName)`, deserializer returned `Ok`";
+
+#[test]
+fn explicit_struct_names() {
+    let options = Options::default().with_default_extension(Extensions::EXPLICIT_STRUCT_NAMES);
+
+    // phase 1 (regular structs)
+    let content_regular = r#"(
+        id: Id(3),
+        position: Position(0.0, 8.72),
+    )"#;
+    let foo = options.from_str::<Foo>(content_regular);
+    assert_eq!(
+        foo.expect_err(EXPECT_ERROR_MESSAGE).code,
+        Error::ExpectedStructName("Foo".to_string())
+    );
+
+    // phase 2 (newtype structs)
+    let content_newtype = r#"Foo(
+        id: (3),
+        position: Position(0.0, 8.72),
+    )"#;
+    let foo = options.from_str::<Foo>(content_newtype);
+    assert_eq!(
+        foo.expect_err(EXPECT_ERROR_MESSAGE).code,
+        Error::ExpectedStructName("Id".to_string())
+    );
+
+    // phase 3 (tuple structs)
+    let content_tuple = r#"Foo(
+        id: Id(3),
+        position: (0.0, 8.72),
+    )"#;
+    let foo = options.from_str::<Foo>(content_tuple);
+    assert_eq!(
+        foo.expect_err(EXPECT_ERROR_MESSAGE).code,
+        Error::ExpectedStructName("Position".to_string())
+    );
```

Comment:
Could you also add a test for this extension combined with unwrap variant newtypes, e.g. that Enum::Variant(Struct { a: 42 }) is still serialised to and deserialised from Variant(a: 42) ? Sorry for only just thinking of this now.

Reply:
consolidated with phase 5 (serialization/deserialization equivalence)